File: cjeongmin/추석_트래픽.js
const parseTimeToMilliseconds = (timeStr) => {
    const [hh, mm, ss] = timeStr.split(":");
    return +hh * 3600000 + +mm * 60000 + parseFloat(ss) * 1000;
};

const parseLog = (line) => {
    const [_, endTimeStr, durationStr] = line.split(" ");
    const endTime = parseTimeToMilliseconds(endTimeStr);
    const duration = parseFloat(durationStr.slice(0, -1)) * 1000;
    const startTime = endTime - duration + 1;

    return [startTime, endTime];
};

const countOverlappingLogs = (timestamp, logs) => {
    const windowEnd = timestamp + 999;
    return logs.filter(([start, end]) => end >= timestamp && start <= windowEnd).length;
};

function solution(lines) {
    const processedLogs = lines.map(parseLog);
    const endTimestamps = processedLogs.map(([_, end]) => end);

    return Math.max(...endTimestamps.map((time) => countOverlappingLogs(time, processedLogs)));
}

File: youngjun/week2/추석_트래픽.py
# answer의 max 값은 len(lines)
# answer는 무조건 특정 로그의 시작 혹은 끝 시점
# 답이 될 수 있는 구간을 intervals 딕셔너리로 만들고, 각 로그에 대해 범위가 겹치는 구간의 카운트를 1씩 올리면 구할 수 있음


def logSplit(lines):
    endTime = []
    duration = []
    for line in lines:
        logs = line.split()
        endTime.append(logs[1])
        duration.append(float(logs[2][:-1]))
    return endTime, duration

def timeToFloat(time):
    h = int(time[0:2])
    m = int(time[3:5]) 
    s = float(time[6:])
    return h*3600 + m*60 + s

def intervalMaker(endTime, duration):
    intervals = {}
    times = []  # (start, end) 저장
    for i in range(len(endTime)):
        end = round(timeToFloat(endTime[i]), 3)
        start = round(end - duration[i] + 0.001, 3)
        times.append((start, end))
        intervals[start] = 0
        intervals[end] = 0
    return intervals, times

def solution(lines):
    endTime, duration = logSplit(lines)
    intervals, times = intervalMaker(endTime, duration)
    for start, end in times:
        for key in intervals.keys():
            if key <= end and key + 1 > start:
                intervals[key] += 1
    return max(intervals.values())